Add Secret and Variable classes

nthings opened the pull request:
Hi! this PR is for adding the Secret and Variable classes. Before I added the variable methods to Organization and Repository but I think is better handling this way. Let me know feedback if this is good so I can start adding test cases. Ty!

trim21 commented:
No need to add `pyi` for new classes, just use inline types.

EnricoMi commented on github/Repository.py:
```diff
@@ -1617,7 +1677,7 @@ def create_variable(self, variable_name: str, value: str) -> bool:
-        return status == 201
+        return github.Variable.Variable(self._requester, headers, data, completed=True)
 
     def update_variable(self, variable_name: str, value: str) -> bool:
```
There should also be an `edit` method in class `Variable` and `Secret`. Then it is arguable to remove this (reduces footprint of Repository class).

EnricoMi commented on github/Repository.py:
```diff
@@ -1599,7 +1629,37 @@ def create_secret(self, secret_name, unencrypted_value):
+        )
 
     def create_variable(self, variable_name: str, value: str) -> bool:
         """
```
Please update the urls, here from

    https://docs.github.com/en/rest/reference/actions/variables#create-a-repository-variable

to

    https://docs.github.com/en/rest/actions/variables#create-a-repository-variable

EnricoMi commented on github/Variable.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        headers, data = self._requester.requestJsonAndCheck("DELETE", self.url)
+
+    def add_repo(self, repo):
```
Maybe note that this only works for organization variables with "selected" visibility, not for repo variables.

EnricoMi replied:
Alternatively, we could have `class OrganizationVariable(Variable)`, which has `add_repo` and `remove_repo`, and `Organization.get_variable` returns `OrganizationVariable`.

EnricoMi commented on github/Organization.py:
```diff
@@ -609,6 +611,36 @@ def create_repo(
+            completed=False,
+        )
+
     def create_secret(
```
Maybe we put `create_secret` before `get_secret` and `get_secrets`.

EnricoMi commented on github/Secret.py:
```diff
@@ -0,0 +1,138 @@
+import github.Repository
+
+
+class Secret(github.GithubObject.CompletableGithubObject):
```
Please see https://github.com/PyGithub/PyGithub/commit/8bdfc3a0d7fc7391ff34e60e20a6987433025417 for some examples how to inline typing into the `.py` file.

EnricoMi commented on github/Secret.py:
```diff
@@ -0,0 +1,138 @@
+            self._updated_at = self._makeDatetimeAttribute(attributes["updated_at"])
+        if "visibility" in attributes:
+            self._visibility = self._makeStringAttribute(attributes["visibility"])
+        if "selected_repositories_url" in attributes:
```
This means we are fetching the list of selected repositories whenever we create a secret instance, no matter if it is used or not. We like to avoid unnecessary API calls.

Further, this list of repos is paginated, so this would fetch only the first 30 repos.

Ideally, `selected_repositories` returns a paginated list of the repos:

    @property
    def selected_repositories(self) -> PaginatedList[Repository]:
        return github.PaginatedList.PaginatedList(
            github.Repository.Repository,
            self._requester,
            self.selected_repositories_url,
            {},
            list_item="repositories",
        )

EnricoMi reviewed (changes requested):
I like having Secret and Variable classes.

nthings commented:
Addressed most of comments can I get a re-review? @EnricoMi @trim21 Also bit of help needed into the implementation of `PaginatedList` on the `selected_repositories` property

trim21 commented on github/OrganizationVariable.py:
```diff
@@ -0,0 +1,138 @@
+        self._selected_repositories.value.remove(repo)
+        return True
+
+    def _initAttributes(self) -> None:
```
put `_initAttributes` before `__repr__`

trim21 commented:
Sorry, I mean move `_initAttributes`, and make it the first method of class, don't move `__repr__`

nthings commented:
> Sorry, I mean to move `_initAttributes`, and make it the first method of class, don't move `__repr__`

Ok. Below the properties right?

trim21 commented:
> > Sorry, I mean to move `_initAttributes`, and make it the first method of class, don't move `__repr__`
> 
> Ok. Below the properties right?

nthings commented:
@trim21 reordered those methods. Do you have example of a class that one of the properties is a PaginatedList?

trim21 commented:
> @trim21 reordered those methods. Do you have example of a class that one of the properties is a PaginatedList?

https://github.com/PyGithub/PyGithub/blob/13e178a3abcdc1acfb3dd63aecc2f5e22b7772ae/github/MainClass.py#L262-L270

nthings commented:
> > @trim21 reordered those methods. Do you have example of a class that one of the properties is a PaginatedList?
> 
> https://github.com/PyGithub/PyGithub/blob/13e178a3abcdc1acfb3dd63aecc2f5e22b7772ae/github/MainClass.py#L262-L270

But if you check the mypy errors and code. How can I add or delete repos from the `selected_repositories` property on the class, that logic worked with simple list before

nthings commented:
@trim21 I think I got it working. Please let me know any feedback so I can go ahead with test cases

EnricoMi reviewed (approved):
LGTM!

Ideally, we would have tests for `get_secret`, `get_secrets`, `create_secrets`, `get_variable`, `get_variables` and `create_variables` on `Organization` and `Repository`, assert all attributes of `Secret`, `OrganizationSecret` and `Variable` (at least in one test), and call `selected_repositories` on `OrganizationSecret` and `OrganizationVariable`.

nthings commented:
@EnricoMi @trim21 Test cases added!

EnricoMi reviewed (approved):
LGTM!
